### PR TITLE
Package ray java jars into wheels

### PR DESCRIPTION
**Pull request by chaokunyang**
## Why are these changes needed?
Package all ray jars into `ray` wheel. Provide a function `get_ray_jars_dir` in `ray.services` to get jars directory in runtime. Ray jars can be used to setup class path for java process, which is necessary for creating java actors in python.

## Related issue number
#6608

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by chaokunyang** [comment]
@kfstorm @zhijunfu

**Comment by robertnishihara** [comment]
@chaokunyang is this intended for Ray Java users? Or is it intended for Ray users that use Python and Java and want to invoke Java from Python?

In terms of making it easier to distribute the Java code, isn't the natural thing to put the jars on maven or something like that? Is this intended to be an alternative to that?

**Comment by chaokunyang** [comment]
@robertnishihara It's for Ray users that use Python and Java and want to invoke Java from Python. For example, When Python driver/workers want to create a Java actor, It needs a way to get Ray java jars. Python process can use jars in wheel to setup java class path and create Java actor. 
For pure Java users, jars will be still distributed to maven and users can use these jars by maven dependencies.

**Comment by chaokunyang** [comment]
In our design of streaming python api, there will be a python driver. The Python driver will create a Java `PythonEntryPoint` actor to map python data stream api call to java, then creates a JobGraph, and submits JobGraph to JobMaster Java actor. JobMaster Java actor will create Python JobWorker according to information of JobGraph. 

So we need to create Java actor from python driver, this wheel package is the way to tell Python driver where ray java jars located.

**Comment by chaokunyang** [comment]
@robertnishihara create an issue(#6608) for this.

**Review by raulchen — COMMENTED** [review]
1) Can you add an assert that checks if `--include-java` is used in `ray start`, `ray-java` package must be installed?
2) Can you remove [this line](https://github.com/ray-project/ray/blob/646643a5884db424bf32604376b7e85a59a23f37/java/test/src/main/java/org/ray/api/test/BaseMultiLanguageTest.java#L87), so the test will use the jars packaged in wheel?

**Comment by chaokunyang** [comment]
Done

**Comment by chaokunyang** [comment]
@raulchen Fixed

File: java/dist/setup.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            return python, abi, plat
+except ImportError:
+    bdist_wheel = None
+
+setup(
+    name="ray-java",
+    packages=["ray_java"],
+    version=find_version(),
+    description="Package all ray java jars into a wheel so that ray can"
+    " create java actor in python code based on jars in the wheel.",
```
[the remark this answers is not in the document]
Comment (chaokunyang) [reply java/dist/setup.py]: That's would be great. Just fixed.

File: python/ray/scripts/scripts.py [hunk 2]
```diff
@@ -46,6 +47,40 @@ def check_no_existing_redis_clients(node_ip_address, redis_client):
                             "clients with this IP address.")
 
 
+def __process_java_worker_classpath(java_options):
```
[the remark this answers is not in the document]
Comment (chaokunyang) [reply python/ray/scripts/scripts.py]: fixed

**Comment by robertnishihara** [comment]
@chaokunyang @raulchen based on our offline conversation, I thought the plan was to put package the jars into the regular Ray wheels instead of having a separate package. E.g., this is what pyspark does I think.

**Comment by raulchen** [comment]
> @chaokunyang @raulchen based on our offline conversation, I thought the plan was to put package the jars into the regular Ray wheels instead of having a separate package. E.g., this is what pyspark does I think.

Yep, I talked to @chaokunyang offline on Friday. We'll try that approach.

**Comment by chaokunyang** [comment]
> @chaokunyang @raulchen based on our offline conversation, I thought the plan was to put package the jars into the regular Ray wheels instead of having a separate package. E.g., this is what pyspark does I think.

Just updated the PR, packaged all ray jars into wheel. Please take a look.

File: java/BUILD.bazel [hunk 3]
```diff
@@ -207,3 +207,28 @@ genrule(
     local = 1,
     tags = ["no-cache"],
 )
+
+java_binary(
```
Comment (raulchen) [inline java/BUILD.bazel]: Can you change the main_class to an empty string, and add this comment:
```
This rule is used to package all Ray Java code and the third-party dependencies into a fat jar file. It's not really an executable jar. So we set its `main_class` to empty.
```

File: python/ray/services.py [hunk 4]
```diff
@@ -1252,6 +1253,17 @@ def start_raylet(redis_address,
...
 
 
+def get_ray_jars_dir():
+    """Return a directory where all ray-related jars and
+      their dependencies locate."""
+    current_dir = os.path.abspath(os.path.dirname(__file__))
+    jars_dir = os.path.abspath(os.path.join(current_dir, "jars"))
+    if not os.path.exists(jars_dir):
+        raise Exception("Ray jars is not packaged into ray. "
+                        "Please build ray with java enabled.")
```
Comment (raulchen) [inline python/ray/services.py]:
```suggestion
                        "Please build ray with java enabled (set env var RAY_INSTALL_JAVA=1")
```

File: java/BUILD.bazel [hunk 5]
```diff
@@ -207,3 +207,28 @@ genrule(
...
+    runtime_deps = [
+        "//java:org_ray_ray_api",
+        "//java:org_ray_ray_runtime",
+        "//streaming/java:org_ray_ray_streaming-api",
+        "//streaming/java:org_ray_ray_streaming-runtime",
+    ]
+)
+
+genrule(
+    name = "ray_java_pkg",
```
Comment (raulchen) [inline java/BUILD.bazel]: we can now change `"$BAZEL_EXECUTABLE" build //java:all --verbose_failures` in `build.sh` to `"$BAZEL_EXECUTABLE" build //java:ray_java_pkg --verbose_failures`.

**Review by raulchen — APPROVED** [review]
LGTM. Only some small comments.

File: java/BUILD.bazel (hunk 3 above)
Comment (chaokunyang) [reply java/BUILD.bazel]: fixed

File: java/BUILD.bazel (hunk 5 above)
Comment (chaokunyang) [reply java/BUILD.bazel]: fixed

Comment (jovany-wang) [reply java/BUILD.bazel]: How would pure Java work here?

Comment (chaokunyang) [reply java/BUILD.bazel]: Please see `java/test.sh`

**Pull request merged**